### PR TITLE
Tell openedx gunicorn to fork workers based on CPU count

### PR DESCRIPTION
**Pull request by silviot**
Tutor currently runs a single worker `gunicorn` process.

This PR makes the number of workers dependent on the number of cores: `(2 x $num_cores) + 1`, as recommended in the [gunicorn documentation](http://docs.gunicorn.org/en/stable/design.html#how-many-workers).

### CODE REVIEW

**Comment by silviot** [comment]
`notes` and `xqueue` will need a similar update. I can work on it, but IMO this PR can already be merged leaving them as TODOs.

File: tutor/templates/build/openedx/gunicorn_conf.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import multiprocessing
+
+# The recommended number of workers is twice the
+# CPU count plus one
+workers = (multiprocessing.cpu_count()-1) * 2 + 2
```
Comment (regisb) [inline tutor/templates/build/openedx/gunicorn_conf.py]: Shouldn't we write instead:

    workers = multiprocessing.cpu_count() * 2 + 1

?

File: tutor/templates/build/openedx/Dockerfile [hunk 2]
```diff
@@ -89,6 +89,9 @@ ENV SETTINGS tutor.production
 # Entrypoint will fix permissions of all files and run commands as openedx
 ENTRYPOINT ["docker-entrypoint.sh"]
 
+# Copy gunicorn settings file
+COPY gunicorn_conf.py /openedx/gunicorn_conf.py
+
 # Run server
 EXPOSE 8000
-CMD gunicorn --name ${SERVICE_VARIANT} --bind=0.0.0.0:8000 --max-requests=1000 ${SERVICE_VARIANT}.wsgi:application
+CMD gunicorn -c file:/openedx/gunicorn_conf.py --name ${SERVICE_VARIANT} --bind=0.0.0.0:8000 --max-requests=1000 ${SERVICE_VARIANT}.wsgi:application
```
Comment (regisb) [inline tutor/templates/build/openedx/Dockerfile]: It's interesting, I didn't know it was possible to use a python configuration file like this. http://docs.gunicorn.org/en/stable/configure.html#configuration-file

I like the fact that we keep the `bind` and `max-requests` parameters out of the configuration file: I think it makes it easier to find the information about the exposed port.

**Comment by regisb** [comment]
thanks @silviot, it's great that you spotted that out! It's crazy that gunicorn works with just a single worker by default: I was absolutely sure that it adapted to the number of cores out of the box.

IMHO xqueue and notes don't need to handle too many requests. I think it's reasonable to set `--workers=3` in those apps. Maybe you can do that in this PR?

File: tutor/templates/build/openedx/gunicorn_conf.py (hunk 1 above)
Comment (silviot) [reply tutor/templates/build/openedx/gunicorn_conf.py]: Oh, yes, we should!

/me hides in shame ;-)

File: tutor/templates/build/openedx/Dockerfile (hunk 2 above)
Comment (silviot) [reply tutor/templates/build/openedx/Dockerfile]: In my opinion a good spot for the `bind` and `max-requests` parameters is the `GUNICORN_CMD_ARGS` environment variable. But we can't do that with the current gunicorn version, so for now I agree they look good in the `CMD` directive.

**Comment by silviot** [comment]
AFAIC this should be good to merge now.

**Comment by regisb** [comment]
thanks again for the contribution!

**Pull request merged**